### DISCUSSION TITLE
[Move OnUpgrade state from Body to Request/Response]

### DISCUSSION THREAD

**Discussion by petar-dambovaliev** [hyperium/hyper#2453] [General]
https://github.com/hyperium/hyper/pull/2353

I have been making a higher level web framework for the past few months.
The mentioned PR breaks my design and requires a lot of work and leads to a lot of headaches.
I was wondering if this "improvement" was worth breaking users.

**Comment by seanmonstar** [comment]
Sorry it's be a headache. I'd also "done the upgrade" in warp, and the changes seemed straightforward, so I didn't think it'd be too bad for most others. As to if it was "worth it", well... Here's a little more motivation behind the change:

- To make the API more consistent with similar-looking extensions that we planned to add, like [Server Push](https://github.com/hyperium/hyper/issues/1586) and [Early Hints](https://github.com/hyperium/hyper/issues/2426). I wouldn't want to keep adding these things to the `Body` type, but if left upgrades there, it'd be inconsistent.
- There's a proposal to make the `Body` type itself [split into more specific variants](https://github.com/hyperium/hyper/issues/2345), and having this `on_upgrade` thing dangling on them would be weirder if we did that.

**Reply by petar-dambovaliev** [reply]
Yea, it was a pain but it took me a day-ish to fix. Thanks for the info.
